Units selector

ErikAmerico opened the pull request:
When a part is loaded into the viewer

The user can select `mm` or `inches` as the unit measurement, updating the db.

If a part already has units defined, those are shown by default and can be still be updated.

If a part hasn't been loaded then the user can't select a unit of measurement and there is a toolTip that explains why they can't view the select dropdown.

ErikAmerico committed:
basic formatting of units selector

ErikAmerico committed:
get part ID in UnitsSelector from ViewPartsButton

ErikAmerico committed:
get part ID in UnitsSelector from uploadPartButton

ErikAmerico committed:
update units in db

ErikAmerico committed:
get and display current units if a file has them defined

ErikAmerico committed:
do not allow a user to select a unit if a part has not been selected. add a toolTip for feedback

rknizzle commented on src/components/AnalysisPage/UnitsSelector/UnitsSelector.jsx:
```diff
@@ -54,20 +54,28 @@ const UnitsSelector = ({ partId }) => {
     }
   };
 
+  const isDisabled = !partId;
+  const tooltipTitle = partId
+    ? ""
+    : "View or upload a part to the viewer before selecting a unit of measurement.";
+
```
This tooltip is great! Is it possible to also make the unit selector less greyed out when there is a part in the page?

I feel like it looks pretty greyed out and unclickable even when their is a part.
![unit-selector-red](https://github.com/rknizzle/propermesh-view/assets/40612461/f3baceb9-65cc-40c4-8243-88c6c038b1c4)



Is that an easy change?

rknizzle replied:
If its not an easy change, then dont worry about this. Its not a big deal

ErikAmerico replied:
@rknizzle yeah it should be an easy change!! 

I'll check it out this evening

ErikAmerico replied:
Just changing the color of the words "select units" to black when they are able to select a measurement cleared that up If you ask me!

rknizzle replied:
Yup this looks awesome :+1:

rknizzle commented on src/components/AnalysisPage/UnitsSelector/UnitsSelector.jsx:
```diff
@@ -26,8 +49,14 @@ const UnitsSelector = ({ partId }) => {
         />
-      </Space>
-    </div>
+        {showSuccess && (
+          <CheckCircleOutlined
+            style={{ color: "green", fontSize: "16px" }}
+            id="unit-update-success-icon"
+          />
+        )}
```
I love this little graphic

rknizzle commented on src/components/AnalysisPage/UnitsSelector/UnitsSelector.jsx:
```diff
@@ -1,20 +1,43 @@
+        setShowSuccess(false);
+      }, 2000);
+    } catch (error) {
+      console.error("Error updating units:", error);
```
It might be good to pop up a little message saying that it failed to update units

ErikAmerico replied:
@rknizzle 
I added a little emoji  in the popup that we do not have to keep! I just wasn't sure what the next step should be for the user after being notified that the units didn't update. 

Should they be able to report an issue somehow?

rknizzle replied:
Yea good question. I just put in a little message saying to try again or report it to my email.
Here are my commits if you're curious: 
[11c9afc](https://github.com/rknizzle/propermesh-view/pull/37/commits/11c9afcbe49423ee8bfc27df83aa385234038da8)
[c1025a3](https://github.com/rknizzle/propermesh-view/pull/37/commits/c1025a3415096f0fa42b2a67d4806a9c3abd7acf)

@ErikAmerico

rknizzle replied:
Also I think just the floating error message is good enough rather than needing to press ok to get out of the modal

rknizzle commented on src/components/AnalysisPage/UnitsSelector/UnitsSelector.jsx:
```diff
@@ -1,20 +1,43 @@
     setUnits(value);
-    console.log(`selected ${value}`);
+    try {
+      const response = await fetch(`/api/v0/parts/${partId}/units`, {
+        method: "PUT",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify({ units: value }),
+      });
+
+      if (!response.ok) {
+        throw new Error("Failed to update units.");
+      }
+
+      const data = await response.json();
+      console.log(data);
```
I find HTTP requests to be good things to put into their own functions. Its a lot of setup code that can make the component harder to read through. I'd recommend putting this into its own function and calling it here.

Code for HTTP requests can also tend to grow a lot over time because code for handling different types of error responses might continuously be added. So its good to have it in its own function to put all of that logic and not have it fill up the React component

ErikAmerico replied:
Makes sense!

I'll follow that practice from now on

rknizzle commented on src/components/AnalysisPage/UnitsSelector/UnitsSelector.jsx:
```diff
@@ -8,6 +8,25 @@ const UnitsSelector = ({ partId }) => {
+    const getCurrentUnits = async () => {
+      if (partId) {
+        try {
+          const response = await fetch(`/api/v0/parts/${partId}`);
+          if (!response.ok) {
+            throw new Error("Failed to get part data.");
+          }
+          const data = await response.json();
```
This can probably go into its own file too. 

I think It'll eventually move up to the AnalysisPage and then just pass the units down to this component.

rknizzle commented on src/components/AnalysisPage/UnitsSelector/UnitsSelector.jsx:
```diff
@@ -8,6 +8,25 @@ const UnitsSelector = ({ partId }) => {
+          const data = await response.json();
+          setUnits(data.units);
+        } catch (error) {
+          console.error("Error getting part data:", error);
```
This would maybe be good to have an error message popup but dont worry about this one right now because I think this call might get moved to a higher component as you start building out the analysis components

rknizzle reviewed (changes requested):
Nice! This works well and I like all of the graphics a lot. 
I left a couple little suggestions in the comments

ErikAmerico committed:
make the placeholder text of unit selector appear more clickable when there is a part without selected measurement

ErikAmerico committed:
add popup that alerts the user when the units failed to update

ErikAmerico committed:
move http call for updating a parts units to own function/file

ErikAmerico committed:
move http call for getting part data into its own function

ErikAmerico committed:
simplify updatePartUnit to not throw redundant console.error messages

Commit:
Update the error modal that pops up when unit update fails

Also keep track of the previous value so that if there is an error it
will switch back to the previous selection and not show the newly
updated selection

Commit:
Update logic for updating the value shown in the selector

In my previous commit I made a change so that if the update fails itll
stay as the same selection as before the selection was made so they know
that the unit change didnt work and this does basically the same thing
and is a lot simpler

rknizzle commented:
Ok this component is looking sick now! Merging this in

ErikAmerico commented:
Cool!!

I originally used the modal as the pop up, because the user had to acknowledge there was an issue by clicking the okay button.

But I like keeping everything as a  notification
